dll/win32/shell32/CDefaultContextMenu.cpp

```diff
@@ -513,53 +539,96 @@ CDefaultContextMenu::AddStaticContextMenusToMenu(
-
-                RegCloseKey(hkVerb);
+                if (LoadStringW(shell32_hInstance, idResource, wszVerb, _countof(wszVerb)))
+                    mii.dwTypeData = wszVerb; /* use translated verb */
+                else
+                    ERR("Failed to load string\n");
             }
+
+            if (RegOpenKeyW(info.hkClass, wszKey, &hkVerb) != ERROR_SUCCESS)
+                hkVerb = NULL;
         }
         else
         {
-            LONG res = RegOpenKeyW(info.hkClass, wszKey, &hkVerb);
-            if (res == ERROR_SUCCESS)
+            if (RegOpenKeyW(info.hkClass, wszKey, &hkVerb) == ERROR_SUCCESS)
             {
-                DWORD cbVerb = sizeof(wszVerb);
-                res = RegLoadMUIStringW(hkVerb, NULL, wszVerb, cbVerb, NULL, 0, NULL);
-                if (res == ERROR_SUCCESS)
+                if (!(uFlags & CMF_OPTIMIZEFORINVOKE))
                 {
-                    /* use description for the menu entry */
-                    mii.dwTypeData = wszVerb;
-                }
+                    DWORD cbVerb = sizeof(wszVerb);
 
-                res = RegQueryValueExW(hkVerb, L"Extended", NULL, NULL, NULL, NULL);
-                Extended = (res == ERROR_SUCCESS);
+                    LONG res = RegLoadMUIStringW(hkVerb, L"MUIVerb", wszVerb, cbVerb, NULL, 0, NULL);
+                    if (res || !*wszVerb)
+                        res = RegLoadMUIStringW(hkVerb, NULL, wszVerb, cbVerb, NULL, 0, NULL);
 
-                RegCloseKey(hkVerb);
+                    if (res == ERROR_SUCCESS && *wszVerb)
+                    {
+                        /* use description for the menu entry */
+                        mii.dwTypeData = wszVerb;
+                    }
+                }
+            }
+            else
+            {
+                hkVerb = NULL;
             }
         }
 
-        if (!Extended || GetAsyncKeyState(VK_SHIFT) < 0)
+        if (hkVerb)
         {
-            mii.cch = wcslen(mii.dwTypeData);
+            // FIXME: GetAsyncKeyState should not be called here, clients
+            // need to be updated to set the CMF_EXTENDEDVERBS flag.
+            if (!(uFlags & CMF_EXTENDEDVERBS) && GetAsyncKeyState(VK_SHIFT) >= 0)
+                hide |= RegValueExists(hkVerb, L"Extended");
+
+            hide |= RegValueExists(hkVerb, L"ProgrammaticAccessOnly");
+
+            if (!(uFlags & CMF_DISABLEDVERBS))
+                hide |= RegValueExists(hkVerb, L"LegacyDisable");
+
+            if (RegValueExists(hkVerb, L"NeverDefault"))
+                fState &= ~MFS_DEFAULT;
+
+            if (RegValueExists(hkVerb, L"SeparatorBefore"))
+                cmdFlags |= ECF_SEPARATORBEFORE;
+            if (RegValueExists(hkVerb, L"SeparatorAfter"))
+                cmdFlags |= ECF_SEPARATORAFTER;
+
+            RegCloseKey(hkVerb);
+        }
+
+        if (((uFlags & CMF_NODEFAULT) && ntver >= _WIN32_WINNT_VISTA) ||
+            ((uFlags & CMF_DONOTPICKDEFAULT) && ntver >= _WIN32_WINNT_WIN7))
```

Comment:
Not required, but you could create a helper function like this to unify both instances of this and make the code slightly easier to read:
```
static inline bool IsNoDefault(UINT uFlags, UINT ntver)
{
    return ((uFlags & CMF_NODEFAULT) && ntver >= _WIN32_WINNT_VISTA) ||
            ((uFlags & CMF_DONOTPICKDEFAULT) && ntver >= _WIN32_WINNT_WIN7);
}
```

Reply:
The thing is, I'm not 100% sure the code should be the same in both places.

Both flags are ignored on XP and I could not determine a significant difference between Vista and later (CMF_DONOTPICKDEFAULT is documented as 7+) but I assume they did not add the flag just for fun.

The documentation for these flags is not stellar.